### PR TITLE
Use ZK as backend storage for deployment state.

### PR DESCRIPTION
**Pull request by obino**
This patch will make zookeeper (instead of the local disk) the persistent
storage for the deployment state. With this change, each AC now reads
routinely the state from ZK: thus we changed the way appservers are
started to make sure the state is propagated to the head node.

The system is now able to better tolerate restarts and failures on nodes:
the main requirement is that the zookeeper cluster is up and running.

### CODE REVIEW

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/1431/

File: AppController/djinn.rb [hunk 1]
```diff
@@ -460,6 +453,10 @@ class Djinn
   LOG_FILE = "/var/log/appscale/controller-17443.log"
 
 
+  # Where to pud the pid of the controller.
```
Comment (menivaitsi) [inline AppController/djinn.rb]: 'pud' -> 'put' ?

File: AppController/djinn.rb [hunk 2]
```diff
@@ -1868,48 +1890,36 @@ def job_start(secret)
...
-        restart_appengine_apps()
-        if my_node.is_login?
-          scale_appservers_within_nodes()
-          scale_appservers_across_nodes()
-        end
-      }
+      @state = "Done starting up AppScale, now in heartbeat mode"
+
+      # Since we now backup state to ZK, don't make everyone do it.
+      # The Shadow has the most up-to-date info, so let it handle this
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Add a period at the end of this sentenve.

File: AppController/djinn.rb [hunk 3]
```diff
@@ -1868,48 +1890,36 @@ def job_start(secret)
...
+      backup_appcontroller_state() if my_node.is_shadow?
+
+      # The following is the core of the duty cycle: start new apps,
+      # restarts apps, terminate non-responsive appserver, and autoscale.
```
Comment (menivaitsi) [inline AppController/djinn.rb]: restarts -> restart

File: AppController/djinn.rb [hunk 4]
```diff
@@ -2685,10 +2680,18 @@ def add_routing_for_appserver(app_id, ip, port, secret)
...
 
-    Djinn.log_debug("Adding AppServer for app #{app_id} at #{ip}:#{port}")
-    @app_info_map[app_id]['appengine'] << "#{ip}:#{port}"
-    # If this function is called twice, ensure there are no duplicate values.
-    @app_info_map[app_id]['appengine'].uniq!()
+    Djinn.log_debug("Adding AppServer for app #{app_id} at #{ip}:#{port}.")
+
+    # Find and remove an entry for this appserver node and app.
+    APPS_LOCK.synchronize {
+      to_add = @app_info_map[app_id]['appengine'].index("#{ip}:-1")
```
Comment (cdonati) [inline AppController/djinn.rb]: Should this be called `to_delete`, or am I missing something?

File: AppController/djinn.rb [hunk 5]
```diff
@@ -1820,21 +1806,54 @@ def job_start(secret)
...
+    HelperFunctions.write_file(PID_FILE, "#{Process.pid}")
+
+    # If we have the ZK_LOCATIONS_FILE, the deployment has already been
+    # configured and started. We need to check if we are a zookeeper host
+    # and start it if needed.
+    if File.exists?(ZK_LOCATIONS_FILE)
+      # We need to check our saved IPs with the list of zookeeper nodes
+      # (IPs can change in cloud environments).
+      begin
+        my_ip = HelperFunctions.read_file("#{APPSCALE_CONFIG_DIR}/my_private_ip").chomp
```
Comment (cdonati) [inline AppController/djinn.rb]: FWIW, this function chomps by default.

File: AppController/djinn.rb [hunk 6]
```diff
@@ -2910,33 +2913,28 @@ def backup_appcontroller_state()
...
-      json_state = restore_from_local_data()
-      if not json_state.empty?
-        Djinn.log_info("Restored data from local data.")
-      end
-    end
-
-    if json_state.empty?
-      Djinn.log_warn("Unable to restore from ZK or local state, not restoring!")
+    if !File.exists?(ZK_LOCATIONS_FILE)
+      Djinn.log_info("#{ZK_LOCATIONS_FILE} doesn't exists: not restoring data.")
```
Comment (menivaitsi) [inline AppController/djinn.rb]: esists -> exist

File: AppController/djinn.rb [hunk 7]
```diff
@@ -4654,20 +4580,19 @@ def start_app_dashboard(login_ip, uaserver_ip)
...
-          'nginx' => 1080,
-          'nginx_https' => 1443,
-          'haproxy' => AppDashboard::PROXY_PORT,
-          'appengine' => ["#{my_private}:8000", "#{my_private}:8001",
-                          "#{my_private}:8002"],
-          'language' => 'python27'
+      AppDashboard.start(login_ip, uaserver_ip, my_public, my_private, @@secret)
+      APPS_LOCK.synchronize {
+        @app_info_map[AppDashboard::APP_NAME] = {
+            'nginx' => 1080,
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Can these ports be defined as constants up at the top?

Comment (menivaitsi) [inline AppController/djinn.rb]: Or in AppDashboard?

File: AppController/djinn.rb [hunk 8]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
     ea.generate(language)
   end
 
-  def start_appengine()
-    @state = "Preparing to run AppEngine apps if needed"
-    Djinn.log_debug("Preparing to run AppEngine apps if needed")
+  # Examine the list of applications to restarts, or the applications that
```
Comment (menivaitsi) [inline AppController/djinn.rb]: restarts -> restart

File: AppController/djinn.rb [hunk 9]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+  # should be running, and respectively restarts or stats them.
```
Comment (menivaitsi) [inline AppController/djinn.rb]: restarts -> restart
stats -> start

File: AppController/djinn.rb [hunk 10]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
-    if @restored == false
-      # Wait till we can talk to the UserAppServer while we restore.
-      uac = UserAppClient.new(my_node.private_ip, @@secret)
-      Djinn.log_info("Need to restore")
-      begin
-        app_list = uac.get_all_apps()
-      rescue FailedNodeException
-        Djinn.log_warn("Failed to get app listing: retrying.")
-        retry
+    # use a copy of @apps_to_restart here since we delete from it in
```
Comment (menivaitsi) [inline AppController/djinn.rb]: use -> Use

File: AppController/djinn.rb [hunk 11]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+    # setup_appengine_application
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Add period at the end.

File: AppController/djinn.rb [hunk 12]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+    apps = @apps_to_restart
+    apps.each { |app_name|
+      if !my_node.is_login?  # this node has the new app - don't erase it here
```
Comment (menivaitsi) [inline AppController/djinn.rb]: this ->  This

Please add a period at the end.

File: AppController/djinn.rb [hunk 13]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
-      app_list.each { |app|
+      Djinn.log_info("About to restart app #{app_name}")
+      APPS_LOCK.synchronize {
+        setup_appengine_application(app_name, "restart")
+        maybe_reload_taskqueue_worker(app_name)
+      }
+    }
+  end
+
+  # Check with the UserAppServer the list of application that should be
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Change this comment to something like:

"Compares the list of applications that should be running according to the UserAppServer with the list we have on the load balancer, and marks the missing apps for start during the next cycle."

File: AppController/djinn.rb [hunk 14]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+  # running, compare them with the list we have on the load balancer, and
+  # marks the missing ones for start at the next cycle.
+  def check_running_apps()
+    # The login node cross-check with the UserAppServer if all the
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Change this comment to something like:

"This operation is only done on the login node."

File: AppController/djinn.rb [hunk 15]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+            Djinn.log_info("Adding #{app} to running apps.")
+
+            # We query the UserAppServer looking for application data, in
+            # particular ports and language.
+            result = uac.get_app_data(app)
+            app_data = JSON.load(result)
+            Djinn.log_debug("Got app data for #{app}: #{app_data}")
+
+            app_language = app_data['language']
+            Djinn.log_info("Restoring app #{app} language #{app_language} with ports #{app_data['hosts']}.")
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Change to something like:

> Djinn.log_info("Restoring app #{app} (language #{app_language}) with ports: #{app_data['hosts']}.")


to improve readability in the output.

File: AppController/djinn.rb [hunk 16]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+        apps_to_load.each { |app|
+          setup_appengine_application(app, "new")
+          maybe_start_taskqueue_worker(app)
+        }
+      }
+    end
 
-          app_language = app_data['language']
-          Djinn.log_info("Restoring app #{app} language #{app_language} with ports #{app_data['hosts']}.")
+    # From here on, we check that the running appservers matches the
```
Comment (menivaitsi) [inline AppController/djinn.rb]: matches -> match

File: AppController/djinn.rb [hunk 17]
```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
...
+    # headnode view.  Only one thread talking to the appmanagerserver at a
```
Comment (menivaitsi) [inline AppController/djinn.rb]: appmanagerserver -> AppManagerServer

File: AppController/djinn.rb [hunk 18]
```diff
@@ -4846,67 +4852,43 @@ def setup_appengine_application(app, state)
...
+      @app_info_map[app]['haproxy'] = find_lowest_free_port(
+        HAProxy::START_PORT)
+    end
+    if @app_info_map[app]['appengine'].nil?
       @app_info_map[app]['appengine'] = []
-    else
-      maybe_reload_taskqueue_worker(app)
     end
+    if !@app_info_map[app]['nginx'] or
+            !@app_info_map[app]['nginx_https'] or
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Please use four space indentation for this continuation line.

File: AppController/lib/zkinterface.rb [hunk 19]
```diff
@@ -114,22 +114,24 @@ def self.init_to_ip(client_ip, ip)
...
 
-
-  # Initializes a new ZooKeeper connection to the "closest" node in the
-  # system. "Closeness" is defined as either "this node" (if it runs
-  # ZooKeeper), or an arbitrary node that runs ZooKeeper. Callers should use
-  # this method when they don't want to determine on their own which
-  # ZooKeeper box to connect to.
-  def self.init(my_node, all_nodes)
-    self.init_to_ip(my_node.private_ip, self.get_zk_location(my_node,
-      all_nodes))
```
Comment (tmarballi) [inline AppController/lib/zkinterface.rb]: Could you please put in comments above the method to explain what it does?

File: AppController/djinn.rb [hunk 20]
```diff
@@ -5234,11 +5148,19 @@ def get_haproxy_stats(app_name)
   # this method reports whether or not AppServers should be added, removed, or
   # if no changes are needed.
   def get_scaling_info_for_app(app_name, update_dashboard=true)
-
     total_requests_seen = 0
     total_req_in_queue = 0
     time_requests_were_seen = 0
 
+    # Let's make sure we have the minimum number of appserver running.
```
Comment (menivaitsi) [inline AppController/djinn.rb]: appserver -> appservers

File: AppController/djinn.rb [hunk 21]
```diff
@@ -5350,15 +5267,15 @@ def try_to_scale_up(app_name)
           # The host needs to have normalized average load less than
           # MAX_LOAD_AVG, current CPU usage less than 90% and enough
           # memory to run an AppServer (and some safe extra).
-          if node['free_memory'] > Integer(@options['max_memory']) + SAFE_MEM and
-            node['cpu'] < MAX_CPU_FOR_APPSERVERS and
-            node['load'] / node['num_cpu'] < MAX_LOAD_AVG
+          if Float(node['free_memory']) > Integer(@options['max_memory']) + SAFE_MEM and
+            Float(node['cpu']) < MAX_CPU_FOR_APPSERVERS and
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Please use 4-space indentation for continuation lines like this one.

File: AppController/djinn.rb [hunk 22]
```diff
@@ -1820,21 +1806,54 @@ def job_start(secret)
...
-      parse_options()
-    else
+    if !restore_appcontroller_state()
       erase_old_data()
       wait_for_data()
-      parse_options()
     end
+    parse_options()
+
+    # We reset the kill received signal since we are starting now.
```
Comment (tmarballi) [inline AppController/djinn.rb]: kill received signal => kill signal received

File: AppController/djinn.rb [hunk 23]
```diff
@@ -5350,15 +5267,15 @@ def try_to_scale_up(app_name)
...
-            node['cpu'] < MAX_CPU_FOR_APPSERVERS and
-            node['load'] / node['num_cpu'] < MAX_LOAD_AVG
+          if Float(node['free_memory']) > Integer(@options['max_memory']) + SAFE_MEM and
+            Float(node['cpu']) < MAX_CPU_FOR_APPSERVERS and
+            Float(node['load']) / Float(node['num_cpu']) < MAX_LOAD_AVG
             # The host can run a new appserver. Let's check if it's a
             # better candidate (we do use lowest cpu load for it).
             if lowest_load == nil or
-              lowest_load > (node['load'] / node['num_cpu'])
+              lowest_load > (Float(node['load']) / Float(node['num_cpu']))
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Same here.

File: AppController/djinn.rb [hunk 24]
```diff
@@ -5424,68 +5331,174 @@ def try_to_scale_down(app_name)
...
-      begin
-        acc.remove_appserver_process(app_name, port)
-      rescue FailedNodeException
-        Djinn.log_warn("Failed to talk to #{appserver_to_use} to remove" +
-          " appserver for app #{app_name}")
+    @app_info_map[app_name]['appengine'].delete("#{appserver_to_use}:#{port}")
+    @last_decision[app_name] = Time.now.to_i
+  end
+
+  # This function unpack an application tarball if needed. A removal of
```
Comment (menivaitsi) [inline AppController/djinn.rb]: unpack -> unpacks

File: AppController/djinn.rb [hunk 25]
```diff
@@ -5424,68 +5331,174 @@ def try_to_scale_down(app_name)
...
+  # the old application code can be forced with a parameter.
+  #
+  # Args:
+  #   app       : the application name to setup
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Please use full sentences for arg descriptions to match the rest of the file.

File: AppController/lib/monit_interface.rb [hunk 26]
```diff
@@ -1,15 +1,25 @@
...
 
 
-# A constant that we use to indicate that we want the output produced
-# by remotely executed SSH commands.
-WANT_OUTPUT = true
+# Where we save the configuration file.
+MONIT_CONFIG = "/etc/monit/conf.d"
+
+
+# Monit is finnicky when it comes to multiple commands at the same time.
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: finnicky -> finicky

File: AppController/lib/monit_interface.rb [hunk 27]
```diff
@@ -43,22 +53,55 @@ def self.start_file(watch, path, action, hours=12)
...
-
     Djinn.log_run("#{MONIT} start -g #{watch}")
   end
 
   def self.restart(watch)
-    Djinn.log_run("#{MONIT} restart -g #{watch}")
+    self.run_cmd("#{MONIT} restart -g #{watch}")
+  end
+
+  # This function unmonitor and optionally stop the service, and remove
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: unmonitor -> unmonitors
stop -> stops
remove -> removes

File: AppController/lib/monit_interface.rb [hunk 28]
```diff
@@ -43,22 +53,55 @@ def self.start_file(watch, path, action, hours=12)
...
+  # the monit configuration file.
+  def self.stop(watch, stop=true)
+    # To make sure the service is stopped, we query monit till the service
+    # is not anylonger running.
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: anylonger is missing a space.

File: AppController/lib/monit_interface.rb [hunk 29]
```diff
@@ -43,22 +53,55 @@ def self.start_file(watch, path, action, hours=12)
...
+        end
+        Djinn.log_debug("Waiting for monit to stop #{watch}.")
+        Kernel.sleep(SMALL_WAIT)
+      }
+    end
+
+    # Now let's find the corresponding configuration file and remove it.
+    config = Dir::glob("#{MONIT_CONFIG}/appscale-#{watch}*")
+    if config.length > 1
+      Djinn.log_info("Found multiple monit config match for #{watch}: #{config}.")
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: match -> matches

File: AppController/lib/monit_interface.rb [hunk 30]
```diff
@@ -105,24 +148,43 @@ def self.write_monit_config(watch, start_cmd, stop_cmd, port,
...
-    Djinn.log_run("#{MONIT} stop -g #{watch}")
+  def self.is_running?(watch)
+    output = self.run_cmd("#{MONIT} summary | grep #{watch} | grep Running")
+    return (not output == "")
   end
 
-  def self.remove(watch)
-    Djinn.log_run("#{MONIT} stop -g #{watch}")
-    Djinn.log_run("#{MONIT} unmonitor -g #{watch}")
+  # This functions returns a list of running applications: the
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: functions -> function

File: AppController/lib/monit_interface.rb [hunk 31]
```diff
@@ -105,24 +148,43 @@ def self.write_monit_config(watch, start_cmd, stop_cmd, port,
...
+  # dev_appservers needs to still be monitored by monit.
+  # Returns:
+  #   A list of application:port
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: Change to something like:
A list of application:port records.

File: AppController/lib/monit_interface.rb [hunk 32]
```diff
@@ -105,24 +148,43 @@ def self.write_monit_config(watch, start_cmd, stop_cmd, port,
...
+  def self.running_appengines()
+    output = self.run_cmd("#{MONIT} summary | grep -E 'app___.*Running'")
+    appengines = output.gsub! /Process 'app___(.*)-([0-9]*).*/, '\1:\2'
+    if appengines
+      Djinn.log_debug("Found these appengines running: #{appengines.gsub("\n", ' ')}.")
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: appengines -> appengine processes

File: AppController/lib/monit_interface.rb [hunk 33]
```diff
@@ -105,24 +148,43 @@ def self.write_monit_config(watch, start_cmd, stop_cmd, port,
...
+      return appengines.split("\n")
+    else
+      Djinn.log_debug("Found no appengines running.")
```
Comment (menivaitsi) [inline AppController/lib/monit_interface.rb]: appengines -> appengine processes

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/1434/

File: AppController/djinn.rb [hunk 34]
```diff
@@ -4705,72 +4633,165 @@ def place_error_app(app_name, err_msg, language)
...
+
+  # Login nodes will compares the list of applications that should be
+  # running according to the UserAppServer with the list we have on the
+  # load balancer, and marks the missing apps for start during the next
+  # cycle.
+  #
+  # All nodes will compares the list of appservers they should be running,
+  # with the list of appservers actually running, and make the necessary
+  # adjustments. Effectively only login nodes and appengine nodes will run
+  # appserver (login nodes runs the dashboard).
```
Comment (cdonati) [inline AppController/djinn.rb]: appserver -> appservers

File: AppController/djinn.rb [hunk 35]
```diff
@@ -4705,72 +4633,165 @@ def place_error_app(app_name, err_msg, language)
...
+              if app_data['hosts'].values[0]['http']
+                @app_info_map[app]['nginx'] = app_data['hosts'].values[0]['http']
+              end
+              if app_data['hosts'].values[0]['https']
+                @app_info_map[app]['nginx_https'] = app_data['hosts'].values[0]['https']
+              end
+            end
+            @app_names = @app_names + [app]
+          rescue FailedNodeEsception
+            Djinn.log_warn("Couldn't check if app #{app} esists on #{db_private_ip}")
```
Comment (cdonati) [inline AppController/djinn.rb]: esists -> exists

File: AppController/djinn.rb [hunk 36]
```diff
@@ -5234,11 +5156,19 @@ def get_haproxy_stats(app_name)
...
   def get_scaling_info_for_app(app_name, update_dashboard=true)
-
     total_requests_seen = 0
     total_req_in_queue = 0
     time_requests_were_seen = 0
 
+    # Let's make sure we have the minimum number of appservers running.
+    Djinn.log_debug("Evaluating app #{app_name} for scaling.")
+    if @app_info_map[app_name]['appengine'].length < @num_appengines
+       Djinn.log_info("App #{app_name} doesn't have appservers.")
```
Comment (cdonati) [inline AppController/djinn.rb]: Should this be "doesn't have _enough_ appservers"?

File: AppController/djinn.rb [hunk 37]
```diff
@@ -5424,68 +5339,174 @@ def try_to_scale_down(app_name)
...
+      "#{@my_private_ip}:#{appengine_port}")
+
+    xmpp_ip = get_login.public_ip
+
+    app_manager = AppManagerClient.new(my_node.private_ip)
+    begin
+      pid = Integer(app_manager.start_app(app, appengine_port,
+        get_load_balancer_ip(), app_language, xmpp_ip,
+        HelperFunctions.get_app_env_vars(app),
+        Integer(@options['max_memory']), get_login.private_ip))
```
Comment (cdonati) [inline AppController/djinn.rb]: This is a bit hard to read. Would  you mind breaking it into 2 statements?

**Comment by cdonati** [comment]
:+1:

**Comment by obino** [comment]
ocd run https://ocd.appscale.com:8080/job/Daily%20Build/1452/

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/1457/

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/1459/

**Pull request merged**